http-clients/src/main/java/com/palantir/remoting3/clients/UserAgent.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting3.clients;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.Lists;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import org.immutables.value.Value;
+
+/**
+ * Constructs, validates, and formats a canonical User-Agent header. Because the http header spec
+ * (https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2) requires headers to be joined on commas, individual
+ * {@link Agent} header strings must never contain commas.
+ */
+@Value.Immutable
+@ImmutablesStyle
+public interface UserAgent {
+    List<Agent> agents();
+
+    @Value.Lazy
+    default String headerFormat() {
+        return Joiner.on(", ").join(Lists.transform(agents(), Agent::headerFormat));
```

Comment:
I think we may want to sort this?

Reply:
The intent here is to add them in the order they were added.  How would you sort them?

Reply:
what's the purpose of having multiple agent headers? one for "http-remoting (3.3.3)" and one for "foo-service/node123 (5.6.7)"? In that case, not sure sorting has a lot of upside / it's unclear how to sort.

Reply:
Sometimes calls go through a layer of libraries and it's useful to have versions of each encoded.  For example: `contour (1.2.3); foundry-spark (4.5.6); foundry-fsclient (7.8.9)`

Reply:
leaving as is